server/modules/playlist_driver.js

```diff
@@ -15,151 +15,181 @@ limitations under the License.
-      ret.nextModule();
-    },
-    playModule(moduleName) {
-      // Force a specific module to play. Now, this particular module doesn't
-      // necessarily exist in any kind of playlist, which presents us with a
-      // choice as to how long to play this module. We'll choose to play it for
-      // as long as the current layout says to play modules.
-      let layout = playlist[layoutIndex];
-
-      // Stop any existing timer so we don't transition early.
-      // TODO(applmak): Consider making the timer management more foolproof by
-      // having the next* or play* methods stop the timer.
-      clearTimeout(timer);
-      // Reset duration for this module.
-      newModuleTime = time.inFuture(layout.moduleDuration * 1000);
-      // Ensure that we won't change layouts until this module is done.
-      newLayoutTime = Math.max(newModuleTime, newLayoutTime);
-      // Now play this module.
-      ret.playModule_(moduleName);
-    },
-    nextLayout() {
-      // Update layoutIndex.
-      layoutIndex = (layoutIndex + 1) % playlist.length;
-      
-      // Show this layout next:
-      let layout = playlist[layoutIndex];
-      
-      // Reset moduleIndex
-      moduleIndex = -1;
-
-      // The time that we'll switch to a new layout.
-      newLayoutTime = time.inFuture(layout.duration * 1000);
-
-      if (monitor.isEnabled()) {
-        monitor.update({playlist: {
-          time: time.now(),
-          event: `change layout`,
-          deadline: newLayoutTime
-        }});
-      }
-
-      debug(`Next Layout: ${layoutIndex}`);
-
-      moduleSM.fadeToBlack(time.now() + 5000).then(() => {
-        // Shuffle the module list:
-        modules = Array.from(layout.modules);
-        random.shuffle(modules);
-
-        Promise.all(layout.modules.map(m => m.whenLoadedPromise)).then(() => ret.nextModule());
-      });
-    },
-    nextModule() {
-      moduleIndex = (moduleIndex + 1) % modules.length;
-
-      // The current layout.
-      let layout = playlist[layoutIndex];
-      
-      // The time that we'll switch to the next module.
-      newModuleTime = time.inFuture(layout.moduleDuration * 1000);
-
-      ret.playModule_(modules[moduleIndex]);
-    },
-    playModule_(module) {
-      // Play a module until the next transition time.
-      moduleSM.playModule(module, time.now());
-
-      if (monitor.isEnabled()) {
-        monitor.update({playlist: {
-          time: time.now(),
-          event: `change module ${module}`,
-          deadline: Math.min(newModuleTime, newLayoutTime)
-        }});
-      }
-
-      // Now, in so many seconds, we'll need to switch to another module 
-      // or another layout. How much time do we have?
-      if (newLayoutTime < newModuleTime) {
-        timer = setTimeout(() => ret.nextLayout(), time.until(newLayoutTime));
-      } else {
-        timer = setTimeout(() => ret.nextModule(), time.until(newModuleTime));
-      }
+class PlaylistDriver {
+  static makeDriver(moduleSM) {
```

Comment:
Why not just use the constructor?

Reply:
This is a static constructor.

Honestly, I think I was going to put the error-listener assignment here, but I didn't. Would you like me to do that here?

Reply:
I'd either put that in there or just drop the static constructor all together and change the caller to ```new PlaylistDriver(moduleSM)```.

Reply:
Sure.